rfcs/text/0005_route_handler.md

```diff
@@ -0,0 +1,184 @@
+```typescript
+interface KibanaResponseToolkit {
+  [method:string]: (...params: any) => KibanaResponse
+}
+router.get(...,
+  (context: Context, request: KibanaRequest, t: KibanaResponseToolkit): KibanaResponse => {
+   return t.ok();
+   // or
+   return t.redirected('/url');
+   // or
+   return t.badRequest(error);
+);
+```
+*KibanaResponseToolkit*  methods allow an end user to adjust the next response parameters:
+- Body. Supported values:`undefined | string | JSONValue | Buffer | Stream`.
+- Status code. 
+- Headers. Supports adjusting [known values](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/v10/http.d.ts#L8) and attaching [custom values as well](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/v10/http.d.ts#L67)
+
+Other response parameters, such as `etag`, `MIME-type`, `bytes` that used in the Legacy platform could be adjusted via Headers.
+
+The router handler doesn't expect that logic inside can throw or return something different from `KibanaResponse`. In this case, Http service will respond with `Server error` to prevent exposure of internal logic details.
+
+#### KibanaResponseToolkit methods
+Basic primitives:
+```typescript
+type HttpResponsePayload = undefined | string | JSONValue | Buffer | Stream;
+interface HttpResponseOptions {
+  headers?: {
+   // list of  known headers
+   ...
+   // for custom headers:
+   [header: string]: number | string | string[] | undefined;
+  }
+}
+
+type StatusCode = 200 | 202 | 204 | 302 | 400 | 401 | 403 | 404 | 409 | 500;
+```
+
+##### Success
+Server indicated that request was accepted:
+```typescript
+type SuccessResponse<T> = <T extends HttpResponsePayload>(
+  payload: T,
+  options?: HttpResponseOptions
+) => KibanaResponse<T>;
+
+const kibanaResponseToolkit = {
+  ok: <T extends HttpResponsePayload>(payload: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(200, payload, options),
+  accepted: <T extends HttpResponsePayload>(payload: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(202, payload, options),
+  noContent: (options?: HttpResponseOptions) => new KibanaResponse(204, undefined, options)
+```
+
+##### Redirection
+The server wants a user to perform additional actions.
+```typescript
+const kibanaResponseToolkit = {
+  redirected: (url: string, options?: HttpResponseOptions) => new KibanaResponse(302, url, options),
+  notModified: (options?: HttpResponseOptions) => new KibanaResponse(304, undefined, options),
+```
+
+#####  Error
+Server signals that request cannot be handled and explains details of the error situation
+```typescript
+// Supports attaching additional data to send to the client
+interface ResponseError extends Error {
+  meta: {
+    data?: JSONValue;
+    code?: string; // error code to simplify search, translations in i18n, etc.
+    link?: string; // link to the docs
```

Comment:
I do too!

Can we make these names more explicit:
- `errorCode` instead of `code` - `code` could mean statusCode, stack trace, etc.
- `docLink` instead of `link` - `link` could mean a link to somewhere in the Kibana application